ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_set(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+		FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+			
+	rb_memcached_return(rc);
 }
 
 static VALUE
-rb_connection_get(VALUE self, VALUE key)
+rb_connection_get(VALUE self, VALUE rb_key)
 {
-  Check_Type(key, T_STRING);
+	VALUE rb_val;
+
+	memcached_st *mc;
+	memcached_return_t rc;
+	size_t ret_len;
+	uint32_t ret_flags;
+	char *ret;
 
-  memcached_ctx *ctx = get_ctx(self);
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
 
-  char *mkey = StringValuePtr(key);
+	ret = memcached_get(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		&ret_len, &ret_flags, &rc);
 
-  size_t return_value_length;
-  memcached_return_t rc;
-  uint32_t flags; // TODO return flags so client can know if it should deserialize or decompress
-  char *response  = memcached_get(ctx->memc, mkey, strlen(mkey), &return_value_length, &flags, &rc);
+	if (ret == NULL) {
+		rb_memcached_error_check(rc);
+		return Qnil;
+	}
 
-  handle_memcached_return(rc);
-  if (rc == MEMCACHED_SUCCESS)
-    return rb_str_new2(response);
-  else
-    return Qnil;
+	rb_val = rb_str_new(ret, ret_len);
+	free(ret);
+
+	/* TODO: return the flags too, maybe */
+	return rb_val;
+}
+
+memcached_return_t
+rb_connection__mget_callback(const memcached_st *mc, memcached_result_st *result, void *context)
+{
+	VALUE rb_result = (VALUE)context;
+	VALUE rb_key = rb_str_new(
+			memcached_result_key_value(result),
+			memcached_result_key_length(result));
+	VALUE rb_value = rb_str_new(
+			memcached_result_value(result),
+			memcached_result_length(result));
+
+	rb_hash_aset(rb_result, rb_key, rb_value);
+	return MEMCACHED_SUCCESS;
 }
 
 static VALUE
 rb_connection_get_multi(VALUE self, VALUE rb_keys)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  VALUE rb_values = rb_hash_new();
-  int i;
-  long keys_len = RARRAY_LEN(rb_keys);
-  const char * keys[keys_len];
-  size_t key_length[keys_len];
+	memcached_execute_fn callbacks[1] = { &rb_connection__mget_callback };
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	long i, keys_len;
+	const char **c_keys;
+	size_t *c_lengths;
+
+	VALUE rb_result;
 
-  VALUE * arr = RARRAY_PTR(rb_keys);
+	UnwrapMemcached(self, mc);
 
-  memcached_return_t rc;
-  memcached_result_st *result;
+	Check_Type(rb_keys, T_ARRAY);
+	keys_len = RARRAY_LEN(rb_keys);
 
-  for(i = 0; i < keys_len; i++) {
-    keys[i] = StringValuePtr(arr[i]);
-    key_length[i] = strlen(keys[i]);
-  }
+	c_keys = alloca(keys_len * sizeof(const char *));
+	c_lengths = alloca(keys_len * sizeof(size_t));
 
-  rc = memcached_mget(ctx->memc, keys, key_length, sizeof(keys)/sizeof(keys[0]));
-  handle_memcached_return(rc);
+	for (i = 0; i < keys_len; i++) {
+		VALUE rb_key = rb_ary_entry(rb_keys, i);
+		Check_Type(rb_key, T_STRING);
 
-  while ((result = memcached_fetch_result(ctx->memc, NULL, &rc))) {
```

Comment:
I removed `memcached_fetch_result`. Looking at the API docs, the alternative `memcached_fetch_execute` is much more performant and clean. The `result` struct it passes to the callback doesn't need to be freed, so the library can reuse it. Check out the `rb_connection__mget_callback` to see how clean it looks now.